### PR TITLE
again adding the ability for spackbot to request maintainers for multiple packages

### PR DESCRIPTION
**Pull request by vsoch**
Second time is the charm? Or maybe I should stop opening so many PRs.

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

### CODE REVIEW

**Comment by vsoch** [comment]
This needs one update to remove requests.

**Comment by vsoch** [comment]
Going to wait for #37 before finishing here, since one of the files has changes we need here.

**Comment by tgamblin** [comment]
@adamjstewart: can you take another look at this one?  I think it looks good but I defer to you on the details, as this is mostly about making life easy for package maintainers like you :).

File: docs/docs/README.md
Comment (adamjstewart) [inline docs/docs/README.md:67]: Is spack-bot a "he"? 😆

File: spackbot/handlers/reviewers.py
Comment (adamjstewart) [inline spackbot/handlers/reviewers.py:105]: I would suggest changing this regex strategy a bit.

1. I would use word breaks instead of spaces so that we can catch titles like:
    ```
    python: build fails for some reason
    ```
    We'll have to see if "word characters" includes dashes or not, but I'm hoping you'll be able to use `\b` here.

2. We may want to consider using fuzzy matching so that things like `PyTorch` match `py-torch`. There have been requests for this on Spack's command-line too. This could be saved for a follow-up PR though.

3. Also, I worry that Spack has so many packages that generic words in issue titles may be matched by accident. For example, `benchmark`, `arm` (and other arches), `gcc` (and other compilers), `bash` (and other shells), `python` (and other languages), `cmake` (and other build systems), and `git` (and other VCS) are all words that commonly appear in issue titles even if the issue has nothing to do with that package. I'm a bit concerned about false positives for these kind of titles. Of course, this fear could be completely unfounded. We could deploy it as is and see how many false positives we get. Or we could require a specific format like:
    ```
    installation issue: <package> .*
    <package> .*
    ```

File: spackbot/handlers/reviewers.py
Comment (adamjstewart) [inline spackbot/handlers/reviewers.py:135]:
```suggestion
                        "- Hey %s, it looks like you might know about the %s package. Can you help with this issue?"
```

**Review by adamjstewart — COMMENTED** [review]
This looks pretty good! I have some concerns about false positives, but this should help a lot. When this is deployed, we may want to remove the section of the issue template that tells people to run `spack maintainers` themselves. Otherwise we'll end up double-pinging maintainers.

File: docs/docs/README.md
Comment (vsoch) [reply docs/docs/README.md:67]: yes.

**Comment by adamjstewart** [comment]
P.S. Love the branch name lol

File: spackbot/handlers/reviewers.py
Comment (vsoch) [reply spackbot/handlers/reviewers.py:105]: That's a good point - I was worried about this too. My thinking is that this is a bit of an experiment in itself, and that we should deploy and then start to keep track of false positives and false negatives, and then adjust accordingly.

**Comment by vsoch** [comment]
https://www.youtube.com/watch?v=rblYSKz_VnI

File: spackbot/handlers/reviewers.py
Comment (adamjstewart) [reply spackbot/handlers/reviewers.py:105]: The only problem is if one or two packages get hit with a lot of false positives and those maintainers get angry at us 😛 

Possible experiment: we could query existing open/closed issue titles and see how many false positives we get. Although it isn't always easy to tell if something is a false positive or not without manually going through all 10K+ titles...

Comment (vsoch) [reply spackbot/handlers/reviewers.py:105]: Oh that sounds fun! yeah I'll totally do that! I have a meeting in 10 mins but will do after.

Comment (vsoch) [reply spackbot/handlers/reviewers.py:105]: okay @adamjstewart here is a first look! I could only run this on about 6K issues because I used up my API limit to get them, lol, but I'll keep updating the data as the limit is restored (I think it's like 5k an hour or something?) Anyway, here is the repo:

https://github.com/vsoch/spacksperiment/tree/main/titles

And here are files of matches to the pattern strategy and misses, respectively:

 - https://github.com/vsoch/spacksperiment/blob/main/titles/matches.json
 - https://github.com/vsoch/spacksperiment/blob/main/titles/misses.json

I think you were spot on about the pattern of a package name followed by a colon - I see this pattern too with dashes so we miss a subset of packages. This could be easily fixed if we replace all special characters with spaces. The other thing I see (which I'm not sure is easy to fix) is the case of a title mentioning multiple packages, or a package in the context of others. E.g., a lot of the mpi ones also mention slurm and vice versa

Comment (vsoch) [reply spackbot/handlers/reviewers.py:105]: I also think we should skip some packages entirely that are just likely to be false positives because they are English words, e.g., "patch"

Comment (vsoch) [reply spackbot/handlers/reviewers.py:108]: okay just tweaked the regex a little more and updated the spacksperiment repo with that, we now have about 12K results (maybe that's about half?)

**Comment by vsoch** [comment]
What specifically are you worried about being broken? It’s a lot of work to split, and very easy to quickly do a PR to disable the endpoint if something is wrong. I would like to review it here and discuss your concerns and get closer to something that you are not worried about.

**Comment by adamjstewart** [comment]
> What specifically are you worried about being broken?

For example, the regex is too permissive and we end up with a lot of false positives.

> It’s a lot of work to split

It seems to me like the two features being added here are completely independent and don't depend on each other in any way. It may be non-negligible work to split them, but they really shouldn't have been in the same PR to begin with.

> very easy to quickly do a PR to disable the endpoint if something is wrong.

That's good, better than reverting.

**Comment by vsoch** [comment]
> For example, the regex is too permissive and we end up with a lot of false positives.

I did an entire experiment repository so that there wouldn't be any doubt about false positives:

https://github.com/vsoch/spacksperiment/tree/main/titles

It's up to 18k, so most of the spack issues that have already existed.

**Comment by vsoch** [comment]
Experiment is updated to 24k titles.

**Comment by tgamblin** [comment]
@vsoch Maybe I'm missing something but how can we check them to understand which ones are false positives and which aren't?  Do we need to read through 24k things?

**Comment by vsoch** [comment]
@tgamblin I was just glancing / scrolling and getting a sense for false positives / negatives - with the recent changes I think it looks pretty good.

**Comment by vsoch** [comment]
okay all issues are added.

**Comment by vsoch** [comment]
Picking up this conversation again - my suggestion is that we flag package names that are also words (e.g., patch) but otherwise the simple regex of package names in titles is a reasonable one. I'm going through package names now to come up with a skip list.

**Comment by vsoch** [comment]
okay I went through the matches.json and here are the list that I think we should add to a "skiplist," meaning removing these terms from the package list:
```
python
less
cuda
stacks (sometimes used to refer to spack stacks, but only 3/7 times)
intel
cmake
mpi
git
file
last
parallel
bash
gcc
c
time
r
root
which
arm
args
acts
mark
patch
perl (?)
java (?)
fortran (?)
check
tree
global
zip
date
scripts
codes
stat
sandbox
```
There were other well known words that could be mistaken, but (for example) terms like "butter" and "octopus" while they might be common in our day to day, it's hugely unlikely someone is going to open an issue with "octopus" in the title and not have it be about the octopus package. So I left those. So what we have is a list of common terms, along with languages/compilers that are likely to show up in a title to reference something other than the package. I put question marks next to the ones I wasn't 100% on. Let me know your thoughts!

**Review by tgamblin — CHANGES_REQUESTED** [review]
Ok so I went through the matches and I guess I have two thoughts: 

1. It looks like the matches are including PR titles as well as issue titles, so it is hard to narrow this down to issues.
2. I still see enough false positives that I am not comfortable flipping the switch on this thing -- it makes me nervous and I think it's going to end up spamming a lot of maintainers. I'm hesitant to scare them off.

Some examples are things like most of the SLURM hits not being things you'd ping the slurm maintainer about -- there are a lot of cases where it's talking about building one of the MPIs with a particular RM variant (`slurm`), and the hit is *only* because of that.  The issue itself is just about `openmpi`, `mpich`, etc. Similarly, most of the CUDA issues are about CUDA variants on other packages, `fftw` has a lot of false positives about variants, etc. There are hits for `perl` that have more to do with its dependents and their interop with it -- same for some other packages.  And `cmake` has tons of hits that are about other packages' builds.

It looks like @vsoch noticed some fo these as they are on her list but I see enough noise in here that I think this is going to rub people the wrong way, and I'm sensitive to that because even if they don't just remove themselves as maintainers, I think people will start ignoring or filtering our messages, which will defeat the purpose of having maintainers.  We just got PR maintainers into Spack and I think it's going well -- people like it -- I don't want to mess with it in quite such an experimental way.

So: can we just add the multiple package request?  I would like to just get that in and hold off on the issue maintainer thing for now.

I did have one thought on how we could handle issues, though: just ask people.  We have the cool new GitHub templates for issues.  They don't ask people at the moment whether their issue has to do with a particular package, but they easily could.  I think we should just add a question there like "does your issue have to do with a particular package?" and have a box for them to put a comma-separated list of spack package names.  Then I would be ok with pinging people, and I think it would make a lot of sense.  That would preserve a fair bit of what's in here but wouldn't be quite as chaotic as I think the autodetection will be.

Thoughts?

**Comment by vsoch** [comment]
Yeah I think that's a good idea, and I'm definitely okay with acting more conservatively. I'll remove the issues check from the PR.

**Comment by tgamblin** [comment]
Thanks @vsoch! I do appreciate all the spackspexperiments!

**Review by tgamblin — APPROVED** [review]

**Pull request merged**
